### PR TITLE
[AMDGPU] Remove SIWholeQuadMode pass early exit

### PR DESCRIPTION
**Pull request by perlfu**
Merge the code bypass elements from the early exit into the main pass execution flow.

This uncovered a bug where lowering of set.inactive operations may not correctly trigger WWM lowering.
Add a trivial optimisation for set.inactive where the value of inactive lanes would be copied from the destination register. This happens when inactive lane value is already computed in WWM. (More obvious when the lowering is happening correctly.)

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp [hunk 1]
```diff
@@ -1493,9 +1496,10 @@ void SIWholeQuadMode::lowerCopyInstrs() {
                                 *MRI, MI->getOperand(0)));
     MI->setDesc(TII->get(CopyOp));
   }
+  return !LowerToCopyInstrs.empty();
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1499]: Just from visual inspection, should this be `!LowerToMovInstrs.empty() || !LowerToCopyInstrs.empty()`?

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp [hunk 2]
```diff
@@ -1688,25 +1686,32 @@ bool SIWholeQuadMode::runOnMachineFunction(MachineFunction &MF) {
...
-  lowerLiveMaskQueries();
-  lowerCopyInstrs();
+  Changed |= lowerLiveMaskQueries();
+  Changed |= lowerCopyInstrs();
 
-  // Shader only needs WQM
-  if (GlobalFlags == StateWQM) {
+  if (!(GlobalFlags & ~StateExact)) {
+    // No wave mode execution
+    Changed |= lowerKillInstrs(false);
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1699]: Before this patch we never called `lowerKillInstrs(false)`. Was that a bug?

**Review by jayfoad — COMMENTED** [review]
Seems reasonable but it would be easier to review if you can separate out the refactoring, the bug fix and the optimization.

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp (hunk 2 above)
Comment (perlfu) [reply llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1699]: No, it wasn't necessary.
We ran the whole lowering for an Exact shader with kills.
In the whole lower `lowerBlock` performs the same function as lowerKillInstrs.

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp (hunk 1 above)
Comment (perlfu) [reply llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1499]: Thanks, I'll fix this.

**Comment by perlfu** [comment]
> Seems reasonable but it would be easier to review if you can separate out the refactoring, the bug fix and the optimization.

I'll separate the patches.
I did some initial compile time testing on an early version of this without the exact mode lowering bypass and the impact was 0.03% increase in compile time -- I think this version is probably lower.

**Comment by perlfu** [comment]
V_SET_INACTIVE fix moved to #98858 and optimizations to #98864.

File: llvm/test/CodeGen/AMDGPU/atomic_optimizations_pixelshader.ll [hunk 3]
```diff
@@ -16,7 +16,8 @@ declare void @llvm.amdgcn.raw.ptr.buffer.store.f32(float, ptr addrspace(8), i32,
 define amdgpu_ps void @add_i32_constant(ptr addrspace(8) inreg %out, ptr addrspace(8) inreg %inout) {
 ; GFX7-LABEL: add_i32_constant:
 ; GFX7:       ; %bb.0: ; %entry
-; GFX7-NEXT:    s_mov_b64 s[10:11], exec
+; GFX7-NEXT:    s_mov_b64 s[8:9], exec
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/atomic_optimizations_pixelshader.ll:19]: Why does this patch cause any codegen changes?

Comment (perlfu) [reply llvm/test/CodeGen/AMDGPU/atomic_optimizations_pixelshader.ll]: Conditions for introducing a variable live mask were slightly different.
I have match them pre-refactor and made them more readable.

**Review by jayfoad — APPROVED** [review]
LGTM, thanks!

**Pull request merged**